libraries/ESP8266MQTTClient/src/ESP8266MQTTClient.cpp

```diff
@@ -0,0 +1,522 @@
+    _connect_info.will_topic = _lwt_topic.c_str();
+    _connect_info.will_message = _lwt_msg.c_str();
+    _connect_info.will_qos = _lwt_qos;
+    _connect_info.will_retain = _lwt_retain;
+
+
+    _connect_info.keepalive = _keepalive;
+    _connect_info.clean_session = clean_session;
+
+
+    _mqtt_state.in_buffer = (uint8_t *)malloc(DEFAULT_MQTT_BUFFER_SIZE_BYTES);
+    _mqtt_state.in_buffer_length = DEFAULT_MQTT_BUFFER_SIZE_BYTES;
+    _mqtt_state.out_buffer =  (uint8_t *)malloc(DEFAULT_MQTT_BUFFER_SIZE_BYTES);
+    _mqtt_state.out_buffer_length = DEFAULT_MQTT_BUFFER_SIZE_BYTES;
+    _mqtt_state.connect_info = &_connect_info;
+
+    mqtt_msg_init(&_mqtt_state.mqtt_connection,
+                  _mqtt_state.out_buffer,
+                  _mqtt_state.out_buffer_length);
+
+    _transportTraits.reset(nullptr);
+    _transportTraits = MQTTTransportTraitsPtr(new MQTTTransportTraits());
+
+    LOG("%s:%s@%s:%d/%s", _username.c_str(), _password.c_str(), _host.c_str(), _port, _client_id.c_str());
+
+    return true;
+}
+
+bool MQTTClient::connected(void)
+{
+    if(_tcp) {
+        return (_tcp->connected() || (_tcp->available() > 0));
+    }
+    return false;
+}
+
+bool MQTTClient::connect(void)
+{
+    int write_len, read_len, connect_rsp_code;
+    if(connected()) {
+        LOG("[MQTT-Client] connect. already connected, try reuse!\n");
+        while(_tcp->available() > 0) {
+            _tcp->read();
+        }
+        return true;
+    }
+
+    if (!_transportTraits) {
+        LOG("[MQTT-Client] connect: MQTTClient::begin was not called or returned error\n");
+        return false;
+    }
+
+    if(_scheme == "mqtt")
+        _tcp = _transportTraits->create();
+    else {
+        LOG("ERROR: currently only support mqtt over tcp\r\n");
+        return false;
+    }
+
+    if(!_tcp->connect(_host.c_str(), _port)) {
+        LOG("[MQTT-Client] failed connect to %s:%u\n", _host.c_str(), _port);
+        return false;
+    }
+
+    LOG("[MQTT-Client] connected to %s:%u\n", _host.c_str(), _port);
+
+    if (!_transportTraits->verify(*_tcp, _host.c_str())) {
+        LOG("[MQTT-Client] transport level verify failed\n");
+        _tcp->stop();
+        return false;
+    }
+
+    _tcp->setNoDelay(true);
+
+    _mqtt_state.outbound_message = mqtt_msg_connect(&_mqtt_state.mqtt_connection,
+                                   _mqtt_state.connect_info);
+    _mqtt_state.pending_msg_type = mqtt_get_type(_mqtt_state.outbound_message->data);
+    _mqtt_state.pending_msg_id = mqtt_get_id(_mqtt_state.outbound_message->data,
+                                 _mqtt_state.outbound_message->length);
+    LOG("Sending MQTT CONNECT message, type: %d, id: %04X\r\n",
+        _mqtt_state.pending_msg_type,
+        _mqtt_state.pending_msg_id);
+    write_len = _tcp->write(_mqtt_state.outbound_message->data,
+                            _mqtt_state.outbound_message->length);
+    LOG("Reading MQTT CONNECT response message\r\n");
+    while (_tcp->available() == 0);
```

Comment:
check if _tcp->connected() while waiting for data, because it will wait forever if the server closes the connection.